core/templates/dev/head/exploration_editor/ExplorationDiffService.js

```diff
@@ -0,0 +1,297 @@
+              v2States[stateData[stateId].newestStateName])) {
+          stateData[stateId].stateProperty = STATE_PROPERTY_UNCHANGED;
+        }
+      }
+
+      // Delete states not present in both v1 and v2
+      for (var stateId in stateData) {
+        if (!v1States.hasOwnProperty(
+              stateData[stateId].originalStateName) &&
+            !v2States.hasOwnProperty(stateData[stateId].newestStateName)) {
+          delete stateData[stateId];
+        }
+      }
+
+      // Track whether terminal nodes in v1 or v2
+      // TODO(bhenning): could show changes to terminal nodes in diff
+      var finalStateIds = [];
+      for (var stateId in stateData) {
+        var oldState = v1States[stateData[stateId].originalStateName];
+        var newState = v2States[stateData[stateId].newestStateName];
+        var oldStateIsTerminal = false;
+        var newStateIsTerminal = false;
+        if (oldState) {
+          oldStateIsTerminal = (
+            oldState.interaction.id &&
+            INTERACTION_SPECS[oldState.interaction.id].is_terminal);
+        }
+        if (newState) {
+          newStateIsTerminal = (
+            newState.interaction.id &&
+            INTERACTION_SPECS[newState.interaction.id].is_terminal);
+        }
+        if (oldStateIsTerminal || newStateIsTerminal) {
+          finalStateIds.push(stateId);
+        }
+      }
+
+      var links = _compareLinks(
+        v1States, originalStateIds, v2States, stateIds);
+
+      return {
+        nodes: stateData,
+        links: links,
+        originalStateIds: originalStateIds,
+        stateIds: stateIds,
+        finalStateIds: finalStateIds
+      };
+    };
+
+    /**
+     * A helper function that takes a list of changes and uses them to modify
+     * stateIds and stateData accordingly.
+     *
+     * stateIds is an object whose keys are the newest state name and whose
+     * values are the assigned state id.
+     * stateData an object whose keys are state IDs and whose value is an object
+     * with these keys:
+     *  - 'newestStateName': the latest name of the state
+     *  - 'originalStateName': the first encountered name for the state
+     *  - 'stateProperty': 'changed', 'unchanged', 'added' or 'deleted'
+     * changeListData is a list of objects, each with two keys:
+     * - changeList: the change list to apply.
+     * - directionForwards: true if changes are compared in increasing version
+     * number, and false if changes are compared in decreasing version number.
+     */
+    var _getDiffGraphData = function(v1States, v2States, changeListData) {
+      var v1Info = _generateInitialStateIdsAndData(v1States);
+      var stateData = v1Info.stateData;
+      var stateIds = v1Info.stateIds;
+      var originalStateIds = angular.copy(stateIds);
+
+      changeListData.forEach(function(changeListDatum) {
+        var changeList = changeListDatum.changeList;
+        var directionForwards = changeListDatum.directionForwards;
```

Comment:
Not here? https://github.com/oppia/oppia/pull/2135/files/6a969180ed94503d2d7c76b717027c3f715970d5#diff-f08ff9d990600e748da105f8518a111eR210